Arm64/Sve: Implement SVE Math *Multiply* APIs

kunalspathak opened the pull request:
- FusedMultiplyAdd
- FusedMultiplyAddBySelectedScalar
- FusedMultiplyAddNegated
- FusedMultiplySubtract
- FusedMultiplySubtractBySelectedScalar
- FusedMultiplySubtractNegated
- MultiplyAdd
- MultiplySubtract
- MultiplyBySelectedScalar

All tests are passing: https://gist.github.com/kunalspathak/511565b3fe4d830dec509d867b8e36b0
Edit: Updated to include `MultiplyAdd` and `MultiplySubtract`

Contributes to https://github.com/dotnet/runtime/issues/99957

kunalspathak committed:
Add *Fused* APIs

kunalspathak committed:
fix an assert in morph

kunalspathak committed:
Map APIs to instructions

kunalspathak committed:
Add test cases

kunalspathak committed:
handle fused* instructions

kunalspathak committed:
jit format

kunalspathak committed:
Added MultiplyAdd/MultiplySubtract

kunalspathak committed:
Add mapping of API to instruction

kunalspathak committed:
Add test cases

kunalspathak commented:
@dotnet/arm64-contrib

a74nh commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1815,6 +1816,22 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
             }
         }
 
+        if ((intrin.id == NI_Sve_FusedMultiplyAddBySelectedScalar) ||
```
Why do these this require special code here?

a74nh commented on src/coreclr/jit/hwintrinsiclistarm64sve.h:
```diff
@@ -46,6 +46,12 @@ HARDWARE_INTRINSIC(Sve,           CreateWhileLessThanOrEqualMask32Bit,
 HARDWARE_INTRINSIC(Sve,           CreateWhileLessThanOrEqualMask64Bit,                              -1,      2,      false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_whilele,    INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_SpecialCodeGen|H ...
 HARDWARE_INTRINSIC(Sve,           CreateWhileLessThanOrEqualMask8Bit,                               -1,      2,      false, {INS_invalid,        INS_sve_whilele,    INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_SpecialCodeGen|H ...
 HARDWARE_INTRINSIC(Sve,           Divide,                                                           -1,      2,      true,  {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_sdiv,       INS_sve_udiv,       INS_sve_sdiv,       INS_sve_udiv,       INS_sve_fdiv,       INS_sve_fdiv},    HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_EmbeddedMaskedOp ...
+HARDWARE_INTRINSIC(Sve,           FusedMultiplyAdd,                                                 -1,     -1,      false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_fmla,       INS_sve_fmla},    HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_EmbeddedMaskedOp ...
```
You are always using `FMLA` for these. Will there be cases where `FMAD` might be more optimal based on register usage? If so, raise an issue to track it.

kunalspathak replied:
Currently, I am just preferencing `op1` as a `targetPrefUse`, in other words telling LSRA to use `op1` as the `targetReg` and mark the registers for other operands as `delayFree`. With that, using `FMLA` will always be optimal. @tannergooding - please correct if I missed anything here.

tannergooding replied:
I would definitely expect us to have some logic around picking FMLA vs FMAD.

The x64 logic is even more complex because it has to handle the RMW consideration (should the `tgtPrefUse` be the addend or multiplier), but it also needs to consider which memory operand should be contained (since it supports embedded loads). That logic is here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L9823 and you'll note that it uses the `node->GetResultOpNumForRmwIntrinsic` to determine which of `op1`, `op2`, or `op3` is both an input and output or otherwise which is last use. It uses this to ensure the right containment choices are being made.

x64 then repeats this logic again in LSRA to actually set the `tgtPrefUse`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lsraxarch.cpp#L2432 and then again in codegen to pick which instruction form it should use: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsiccodegenxarch.cpp#L2947

I expect that Arm64 just needs to mirror the LSRA and codegen logic (ignoring any bits relevant to containment) and picking `FMLA` vs `FMAD` (rather than `231` vs `213`, respectively)

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -560,7 +568,50 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    }
+                    else
+                    {
+                        // If the instruction just has "predicated" version, then move the "embMaskOp1Reg"
+                        // into targetReg. Next, do the predicated operation on the targetReg and last,
+                        // use "sel" to select the active lanes based on mask, and set inactive lanes
+                        // to falseReg.
+
+                        assert(HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrinEmbMask.id));
+
+                        if (targetReg != embMaskOp1Reg)
+                        {
+                            GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, embMaskOp1Reg);
+                        }
+
+                        GetEmitter()->emitIns_R_R_R_R(insEmbMask, emitSize, targetReg, maskReg, embMaskOp2Reg,
+                                                      embMaskOp3Reg, opt);
+
+                        GetEmitter()->emitIns_R_R_R_R(INS_sve_sel, emitSize, targetReg, maskReg, targetReg, falseReg,
+                                                      opt, INS_SCALABLE_OPTS_UNPREDICATED);
```
Is there an assumption being made about the instruction being RMW here?

`FMLA` encodes 4 registers (`Zda`, `Pg`, `Zn`, and `Zm`) where `Zda` is both the source and destination and the operation is functionally similar to `Zda += (Zn * Zm)` (with only a single rounding operation).

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), Zda)` it can then be encoded as simply:
```asm
fmla Zda, Pg/M, Zn, Zm
```

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge)` it can then be encoded as simply:
```asm
movprfx Zda, Pg/M, merge
fmla Zda, Pg/M, Zn, Zm
```

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), Zero)` it can then be encoded as simply:
```asm
movprfx Zda, Pg/Z, Zda
fmla Zda, Pg/M, Zn, Zm
```

There are then similar versions possible using `fmad` when the multiplier is the source and destination (`op2Reg == tgtReg` or `op3Reg == tgtReg`).

-----------------

We should actually never need `sel` for this case, but only need complex generation if `tgtReg` is unique from all input registers (including the merge) and we're merging with a non-zero value, such as `dest = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge)`:
```asm
mov dest, Zda
movprfx dest, Pg/M, merge
fmla dest, Pg/M, Zn, Zm
```

This ends up being different from the other fallbacks that do use `sel` specifically because it's `RMW` and requires predication (that is there is no `fmla (unpredicated)`).

tannergooding replied:
The main goal of using `ins (unpredicated); sel` in the other case is because it allows a 2 instruction sequence as the worst case.

In this case, we at worst need a 3 instruction sequence due to the required predication on the instruction. Thus, it becomes better to use `mov; movprfx (predicated); ins (predicated)` instead as it can allow `mov` to be elided by the register renamer.

kunalspathak replied:
>  such as dest = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge):

For the similar reasoning mentioned in https://github.com/dotnet/runtime/pull/100743#discussion_r1577289650 (where we should only `movprfx` the inactive lanes from `merge -> dest`, the code should be:


```asm
mov dest, Zda
fmla dest, Pg/M, Zn, Zm
sel dest, Pg/M, dest, merge

```

kunalspathak replied:
Actually, I misinterpreted the value of `Pg/M` as AllTrue. Spoke to @tannergooding offline and we would like to generate:

```asm
sel dest, Pg/M, Zda, merge
fmla dest, Pg/M, Zn, Zm
```

kunalspathak committed:
Handle mov Z, Z instruction

kunalspathak committed:
Reuse GetResultOpNumForRmwIntrinsic() for arm64

kunalspathak committed:
Reuse HW_Flag_FmaIntrinsic for arm64

kunalspathak committed:
Mark FMA APIs as HW_Flag_FmaIntrinsic

kunalspathak committed:
Handle FMA in LSRA and codegen

kunalspathak committed:
Remove the SpecialCodeGen flag from selectedScalar

kunalspathak committed:
address some more scenarios

kunalspathak committed:
jit format

kunalspathak committed:
Add MultiplyBySelectedScalar

kunalspathak committed:
Map the API to the instruction

kunalspathak committed:
fix a bug where *Indexed API used with ConditionalSelect were failing

` Sve.ConditionalSelect(op1, Sve.MultiplyBySelectedScalar(op1, op2, 0), op3);` was failing
because we were trying to check if `MultiplyBySelectedScalar` is contained and we hit the assert
because it is not containable. Added the check.

kunalspathak committed:
unpredicated movprfx should not send opt

kunalspathak committed:
Add the missing flags for Subtract/Multiply

kunalspathak committed:
Added tests for MultiplyBySelectedScalar

Also updated *SelectedScalar* tests for ConditionalSelect

kunalspathak committed:
fixes to test cases

kunalspathak committed:
fix the parameter for selectedScalar test

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into sve_math6

kunalspathak committed:
jit format

tannergooding commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -4250,9 +4250,11 @@ void emitter::emitIns_Mov(
         case INS_sve_mov:
         {
-            if (isPredicateRegister(dstReg) && isPredicateRegister(srcReg))
+            // TODO-SVE: Remove check for insOptsNone() when predicate registers
+            //  are present.
```
Do we know what's causing this new TODO?

kunalspathak replied:
Since predicate registers are alias of vector registers, it would just match the first condition, regardless of what type of registers. We have similar TODO at other places that will be removed once we add proper predicate registers.

tannergooding replied:
Ah, I see. That's definitely unfortunate. I'd guess it will also become a real problem once we get into the instructions that move/convert between vector and mask registers.

kunalspathak replied:
IIRC, we have TODO listed at various places to handle it and we have added special `insOpts` for them, which will be removed once predicate register support is added.

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -559,7 +567,117 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    assert(falseReg != embMaskOp3Reg);
+
+                    bool useAddend = true;
+                    if (targetReg == embMaskOp2Reg)
+                    {
+                        useAddend = false;
+                        std::swap(embMaskOp1Reg, embMaskOp2Reg);
+                    }
+                    else if (targetReg == embMaskOp3Reg)
+                    {
+                        useAddend = false;
+                        std::swap(embMaskOp1Reg, embMaskOp3Reg);
+                    }
```
nit: These seem like they could do with a comment as to why its safe/correct to swap. -- I'm not sure they actually are correct as is.

In particular the intrinsic is:
* `op1 == addend`
* `op2 == multiplicand`
* `op3 == multiplier`

Then we have:
* FMLA is `reg1 = reg1 + (reg2 * reg3)`
* FMAD is `reg1 = (reg1 * reg2) + reg3`

Given that:
* when `targetReg == op1Reg`, then things are already fine and need no sorting, we will emit `FMLA`
* when `targetReg == op2Reg`, then we need to swap op2 and op3, then op1 and op3, allowing us to emit `FMAD`
  * if we just swap `op1` and `op2` like you're doing, the addend ends in `reg2` and the result will be wrong
* when `targetReg == op3Reg`, then we need to swap op1 and op3, allowing us to emit `FMAD`

So I think this rather needs to be:
```suggestion
                     if (targetReg == embMaskOp2Reg)
                    {
                        useAddend = false;
                        std::swap(embMaskOp2Reg, embMaskOp3Reg);
                        std::swap(embMaskOp1Reg, embMaskOp3Reg);
                    }
                    else if (targetReg == embMaskOp3Reg)
                    {
                        useAddend = false;
                        std::swap(embMaskOp1Reg, embMaskOp3Reg);
                    }
```

kunalspathak replied:
ah yes, I missed a case for `targetReg == embMaskOp2Reg`. Will fix it along with the comments.

kunalspathak replied:
For `targetReg == op2Reg` case, alternatively we can also have:
1. `reg2 = (reg2 * reg3) + reg1`
2. `reg2 = (reg2 * reg1) + reg3`  ; swap `reg1 <--> reg3`
3. `reg1 = (reg1 * reg2) + reg3` ; swap `reg1 <--> reg2`
4. Use `FMAD`

Not that one is better than other, but just want to make sure that I am getting the calculations right.

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -559,7 +567,117 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                        //      movprfx target, P/m, embMaskOp1Reg
+                        //      fmla    target, P/m, embMaskOp2Reg, embMaskOp3Reg
+                        //
+                        // targetReg != falseReg && targetReg == embMaskOp1Reg
+                        //      sel     target, P/m, embMaskOp1Reg, falseReg
+                        //      fmla    target, P/m, embMaskOp2Reg, embMaskOp3Reg
```
Can't this one be `movprfx, fmla` since the targetReg is `embMaskOp1Reg`?

That should make it safe to do `movprfx embMaskOp1Reg, P/M, falseReg` (or `movprfx embMaskOp1Reg, P/z` in the case that `falseReg` is known to be `VectorZero`) and then the actual `fmla` which updates the active elements, keeping everything correct

kunalspathak replied:
> That should make it safe to do movprfx embMaskOp1Reg, P/M, falseReg

Not sure if I follow, but won't that overwrite the `embMaskOp1Reg` of active lanes (on which we want to do an operation) with `falseReg`?

We want something like this:

```asm
not p6, p7
movprfx embMaskOp1Reg p6, falseReg
fmla target, p7, embMaskOp2Reg, embMaskOp3Reg
```

tannergooding replied:
Oh right. I keep getting this mixed up... I intuitively expect `movprfx` to support setting the inactive elements so that this behavior all works as expected and so that we can always emit a most optimal 2 instruction sequence.

But in practice it instead only sets the active elements so this doesn't work for cases like this and we require `sel; fmla` instead :/

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -559,7 +567,117 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                        //      sel     target, P/m, embMaskOp1Reg, falseReg
+                        //      fmla    target, P/m, embMaskOp2Reg, embMaskOp3Reg
+                        //
+                        // Note that, we just check if the targetReg/falseReg or targetReg/embMaskOp1Reg
+                        // coincides or not. Other combination like falseReg/embMaskOp*Reg cannot happen
+                        // because we marked embMaskOp*Reg as delayFree.
```
Can we assert this somewhere?

kunalspathak replied:
yes, the asserts are added at the top in https://github.com/dotnet/runtime/pull/102007/files#diff-45c9473f1b1d031652a13186a211c1aef1406a74f576d5f04e3a65766b0f5050R575-R577

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -626,6 +744,12 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                                                       INS_SCALABLE_OPTS_UNPREDICATED);
                     }
                     break;
+                case 4:
```
What code is hitting this path?

tannergooding replied:
Is it the `*BySelectedIndex` variants?

kunalspathak replied:
I think I added it here, but forgot to remove it when I moved the code above. Will remove it.

tannergooding replied:
👍, was asking since the assert is saying it shouldn't be RMW, but all the FMA instructions are

kunalspathak committed:
Contain(op3) of CndSel if op1 is AllTrueMask

kunalspathak committed:
Handle FMA properly

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -559,7 +567,166 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                            unreached();
+                    }
 
+                    if (intrin.op3->IsVectorZero())
```
Should this be asserting that `intrin.op3` is contained?

kunalspathak replied:
added `

kunalspathak committed:
added assert
